### DISCUSSION TITLE
How does it magically use {} or [] for arrays?

### DISCUSSION THREAD

**Discussion by alexgleason** [porsager/postgres#939] [Q&A] [answered]
Hi @porsager I've been studying this code for about 6 hours and I have a massive headache. Please put me out of my misery and tell me how you are doing this magic.

Somehow this library can encode jsonb arrays and native arrays from JS arrays perfectly depending on the context. For example, here I am inserting data into a jsonb column and text[] column:

```ts
sql`insert into users (data, arr) values (${['yolo', 'gg']}, ${['hello', 'world']})`
```

which results in this:

```
LOG:  execute uaxwq8akli5: insert into users (data, arr) values ($1, $2)
DETAIL:  parameters: $1 = '["yolo", "gg"]', $2 = '{hello,world}'
```

Even though I used a JS array for both parameters in my code, this library magically converts the second parameter to a native array while the first one uses jsonb.

HOW?

I am trying to get pglite working properly but running into this issue. https://github.com/electric-sql/pglite/issues/220#issuecomment-2334377965 Please end my suffering.

**Comment by boromisp** [comment, marked as answer]
https://www.postgresql.org/docs/current/protocol-flow.html#PROTOCOL-FLOW-EXT-QUERY
https://www.postgresql.org/docs/current/protocol-message-formats.html

If you wait for the `ParameterDescription` message before issuing `Bind`, you can use that to decide how to serialize you parameters.

**Reply by alexgleason** [reply]
Thank you very much! Mystery solved! :tada: :brain:
